src/runtime_src/core/tools/xbmgmt2/OO_NIFD.cpp

```diff
@@ -19,24 +19,116 @@
 #include "tools/common/XBUtilities.h"
 namespace XBU = XBUtilities;
 
+#include "core/common/device.h"
+#include "core/common/system.h"
+#include "core/common/error.h"
+#include "core/common/utils.h"
+
 // 3rd Party Library - Include Files
+#include <boost/format.hpp>
+#include <boost/program_options.hpp>
+namespace po = boost::program_options;
 
 // System - Include Files
 #include <iostream>
+#include <vector>
+#include <fstream>
+#include <fcntl.h>
+
+
+// =============================================================================
+
+// ------ L O C A L   F U N C T I O N S ---------------------------------------
+
+namespace {
+static void
+device_status(const std::string& device) 
+{
+  auto dev = xrt_core::get_mgmtpf_device(xrt_core::utils::bdf2index(device));
+  xrt_core::scope_value_guard<int, std::function<void()>> fd { 0, nullptr };
+  try {
+      fd = dev->file_open("nifd_pri", O_RDWR); 
+  } catch (const std::exception& e) {
+      xrt_core::send_exception_message(e.what(), "XBMGMT");
+  }
+
+  // const int NIFD_CHECK_STATUS = 8;
+  unsigned int statusReg = 0;
+  int result = 0;//dev->ioctl(fd.get(), NIFD_CHECK_STATUS, &statusReg);
+  if (result != 0)
+    throw xrt_core::error("Could not read status register");
+  else
+    std::cout << boost::format("Current NIFD status: 0x%x\n") % statusReg;
+}
+
+static void
+readback(const std::string& device, const std::string file)
+{
+  std::ifstream fin(file.c_str()) ;
+  if (!fin)
+    throw xrt_core::error(boost::str(boost::format("Could not open %s for reading") % file));
+
+  std::vector<unsigned int> hardwareFramesAndOffsets;
+  while (!fin.eof())
+  {
+    unsigned int nextNum = 0 ;
+    fin >> std::ws >> nextNum >> std::ws ;
```

Comment:
This code assumes a given format, what happens if the format changes :(  
Can it be updated to detect such a condition.

Reply:
I don't completely understand what this code does. This is maintained by XDP. So the plan is to get this skeleton code in,  work with Jason to figure out what else he requires for this code to work. As for the formatting, I'm guessing we can fix it when we see discrepancies in the outputs? I can't run the code to make changes and test the output...